impulse/lfsr.nim

```diff
@@ -0,0 +1,234 @@
+##
+## LFSRs used in many digital communication systems (including, for example LTE
+## and 5GNR). For more information see:
+## https://simple.wikipedia.org/wiki/Linear-feedback_shift_register
+##
+## Notes:
+## - This code is heavily based on Nikesh Bajaj's pylfsr, which can be
+## found in https://pylfsr.github.io
+## - This implementation is not optimized for performance, but for simplicity.
+##   It would be relatively trivial to implement a much faster version by
+##   operating on the bits directly, rather than using tensors.
+
+import arraymancer
+import std / [algorithm, strformat]
+
+type LFSR_TYPE* = enum
+  ## LFSR types (see https://simple.wikipedia.org/wiki/Linear-feedback_shift_register)
+  fibonacci, galois
+
+type LFSR_INIT_STATE_TYPE* = enum
+  ## Common LFSR Initial States
+  ## While it is possible to set any LFSR initial `state` by passing a
+  ## Tensor[bool] to `initLFSR`, for convenience it is also possible
+  ## to pass one of these enum values as the initial `state`:
+  ## - `single_true`: all bits set to `false` except the LSB (i.e. the last one)
+  ## - `all_true`: all bits set to `true`
+  single_true, all_true
+
+type LFSR* = object
+  ## LFSR object used to generate fibonacci or galois pseudo random sequences
+  ## To use it first create the object using `initLFSR` and then call either
+  ## the `next` procedure to get the sequence values one by one or `generate`
+  ## to generate multiple values in one go.
+  taps*: Tensor[int]
+  conf*: LFSR_TYPE
+  state*: Tensor[bool]
+  init_state: Tensor[bool]
+  verbose*: bool
+  counter_starts_at_zero*: bool
+  outbit*: bool
+  count*: int
+  seq_bit_index: int
+  sequence: Tensor[bool]
+  feedbackbit: bool
+
+proc initLFSR*(taps: Tensor[int] | seq[int],
+    conf = fibonacci,
+    state: Tensor[bool],
+    verbose = false,
+    counter_starts_at_zero = true): LFSR =
+  ## Initialize LFSR with given feedback polynomial and initial state
+  ##
+  ## Inputs:
+  ## - taps: Feedback polynomial taps as a Tensor[int] or seq[int] of exponents
+  ##         in descending order. Exponent 0 can be omitted, since it is always
+  ##         implicitly added. For example, to use the `x^5 + x^3 + 1`
+  ##         polynomial you can set taps to `[5, 3]` or to `[5, 3, 0]` (or to
+  ##         `[5, 3].toTensor` or `[5, 3, 0].toTensor`).
+  ## - state: Initial state of the LFSR as a Tensor[bool] of size equal to the
+  ##          highest exponent in taps (which must be its first element)
+  ## - conf: LFSR type as an enum (`fibonacci` or `galois`)
+  ## - verbose: Enable it to print additional logs
+  ## - counter_starts_at_zero: Start the count from 0 or 1 (defaults to `true`)
+  ##
+  ## Return:
+  ## - Ready to use LFSR object
+
+  # Remove the last value from taps if it is a zero
+  when typeof(taps) is Tensor:
+    let taps = if taps[taps.size - 1] == 0: taps[_..^2] else: taps
+  else:
+    let taps = if taps[taps.size - 1] == 0: taps.toTensor[_..^2] else: taps.toTensor
```

Comment:
Why not just:

```suggestion
  var taps = when typeof(taps) is Tensor: taps else: taps.toTensor
  if taps[taps.size - 1] == 0: 
    taps = taps[_..^2]
```

The `taps.size` in the `else` branch would likely fail, or did we ever add a `size` for `seq` as compatibility with `Tensor`?

(could be `T` if it was an explicit generic of course. Not really important, just noticed the `.size` and thought one could remove most 'logic' from the equivalent branches).

Reply:
Umm, I agree. It's odd because I added a test in which the taps are set as a seq.
In any case your version is much better. I didn't know you could use a when in this way.